Test/Core/Utility/CoroutinesTest.cs

```diff
@@ -0,0 +1,195 @@
+
+			// All code until first yield is already executed
+			Assert.AreEqual(10, this.x);
+			Assert.True(coroutine.Status == CoroutineStatus.Running);
+
+			scene.Update();
+
+			// All code until second yield is executed
+			Assert.AreEqual(20, this.x);
+			Assert.True(coroutine.Status == CoroutineStatus.Running);
+
+			coroutine.Dispose();
+			scene.Update();
+
+			// Coroutine disposed
+			Assert.AreEqual(20, this.x);
+			Assert.True(coroutine.Status == CoroutineStatus.Disposed);
+
+			scene.Update();
+
+			// No further changes
+			Assert.AreEqual(20, this.x);
+			Assert.True(coroutine.Status == CoroutineStatus.Disposed);
+		}
+
+		[Test] public void Resuming()
+		{
+			Scene scene = new Scene();
+			scene.Activate();
+
+			Coroutine coroutine = Coroutine.Start(scene, this.BasicRoutine(), "resuming");
+
+			// All code until first yield is already executed
+			Assert.AreEqual(10, this.x);
+			Assert.True(coroutine.Status == CoroutineStatus.Running);
+
+			coroutine.Pause();
+			scene.Update();
+
+			// No changes, since the coroutine is now paused
+			Assert.AreEqual(10, this.x);
+			Assert.True(coroutine.Status == CoroutineStatus.Paused);
+
+			scene.Update();
+			scene.Update();
+			scene.Update();
+
+			// No matter how many updates
+			Assert.AreEqual(10, this.x);
+			Assert.True(coroutine.Status == CoroutineStatus.Paused);
+
+			coroutine.Resume();
+			scene.Update();
+
+			// All code until second yield is executed
+			Assert.AreEqual(20, this.x);
+			Assert.True(coroutine.Status == CoroutineStatus.Running);
+		}
+
+		[Test] public void WaitTwoSeconds()
+		{
+			Scene scene = new Scene();
+			scene.Activate();
+
+			int secondsToWait = 2;
+			Coroutine coroutine = Coroutine.Start(scene, this.WaitSeconds(secondsToWait), "waiting");
+
+			// All code until first yield is already executed
+			Assert.AreEqual(10, this.x);
+			TimeSpan startTime = Time.GameTimer;
+
+			// Waiting...
+			while (this.x == 10)
+			{
+				scene.Update();
+				Time.FrameTick(true, true);
+			}
+
+			double secondsElapsed = (Time.GameTimer - startTime).TotalSeconds;
+
+			Assert.AreEqual(20, this.x);
+
+			// Not being overly precise about the 2 seconds wait..
+			Assert.GreaterOrEqual(secondsElapsed, secondsToWait);
+			Assert.LessOrEqual(secondsElapsed, secondsToWait + Time.SecondsPerFrame * 10);
+
+			Assert.True(coroutine.Status == CoroutineStatus.Complete);
+		}
+
+		[Test] public void Exception()
+		{
+			Scene scene = new Scene();
+			scene.Activate();
+
+			Coroutine coroutine = Coroutine.Start(scene, this.ExceptionRoutine(), "exception");
+
+			// All code until first yield is already executed
+			Assert.True(coroutine.Status == CoroutineStatus.Running);
+
+			scene.Update();
+			// All code until second yield is executed
+			Assert.True(coroutine.Status == CoroutineStatus.Running);
+
+			// Exception thrown, a log should appear
+			scene.Update();
+			Assert.True(coroutine.Status == CoroutineStatus.Error);
+		}
+
+		private int x = 0;
```

Comment:
We should avoid shared state, its better to pass in a object that holds this value.

With a unit test runner that runs tests in parallel this might even break randomly.

Reply:
Done